### ISSUE TITLE
Add missing function declaration code action does not appear when there is similar name in the file

### ISSUE THREAD

**Issue by KostkaBrukowa** [microsoft/TypeScript#51891]
# Bug Report
"Add missing function declaration" does not appear for functions that have similar name to other variables in the file. I think that this code action should appear on every function that is not declared nor imported even if there are names similar to it.

### 🔎 Search Terms

Code action
Missing function declaration

### 🕗 Version & Regression Information

It is present since the appearance of this code action in typescript.

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about code actions

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/PTAEDMHsCdQFwBYEsDOoCuA7AJgU3EprthFgMZxKSYA0oZkeoAhhVZqABTPYkC2qFIQDmpTG2qg8ZADbNozStQCUoVKAAO0XClyY4AKHDklmACoJFAEUg6AcpDgBRAB6o4nZQZCgAeqABaenRoFBhQBFxtA28wACN0OAhwxHVqXHhI7TU0ahkAT1AAIjJLTGEMlA1cGRkReEgAOmai+kYM1lMDBkwUJMxHXDc+vSSAXlAB52G4Uc8AbljQZZXVtfXl-yCyELDYLNwYoA)

### 💻 Code

```ts
// for this undefined function, code action (add missing function declaration) is present
functionThatDoesNotExist()
// ^ - cursor here

// but for this one there is only "change spelling to..." code action
const notexistent = notExistent();
//                       ^ - cursor here
```

### 🙁 Actual behavior

"Add missing function declaration" code action is not present for second function

### 🙂 Expected behavior

"Add missing function declaration" code action should be present for second function

**Comment by RyanCavanaugh** [comment]
Some people have a key binding to fix things if and only if there's a single action available, so this would break their workflow. I'd like to get more feedback on this before changing it -- reducing visual noise is always a plus, and this obviously isn't a blocker.

**Comment by nihirv** [comment]
Agree with OP. There's no reason "add missing function declaration" should be missing from the quick fix just because a function name has a similar name to a variable

**Comment by KostkaBrukowa** [comment]
Personally I think it's hard to have a workflow around those features because:
1. it's really hard to predict which code action will be there. You would have to know (somehow) if similar name exists somewhere in the file which in larger files is practically impossible.
2. `change spelling` is not so smart - it does not check types or anything, it just tells you that there is similar name in the file - in my example TS suggests to rename to variable that is not even defined yet (and I've checked that even if the variable would be defined and it's type would be e.g. string it would still suggest renaming even if the variable needs to be a function). Maybe if `change spelling` would be smarter and would be based on actual types of variable the case would be different.

I think that adding missing declaration should have higher priority than change spelling because it's more predictable and applies to more cases.

**Comment by pokey** [comment]
I also agree with OP.  I rely pretty heavily on being able to ask Typescript to write function definitions for me, and it's quite painful when that quick fix action disappears.  

Can you elaborate a bit on your concern wrt someone expecting there to only be one action?  Maybe you could mark the spelling fix as `isPreferred`, so that then they could map something like


```json
{
  "key": "shift+ctrl+e",
  "command": "editor.action.codeAction",
  "args": {
    "kind": "...",
    "preferred": true,
    "apply": "ifsingle"
  }
}
```

**Comment by PixelatedPope** [comment]
Just want to throw my mitre in the discussion. I have found a workaround though. 
When I run into this:

<img width="326" height="189" alt="Image" src="https://github.com/user-attachments/assets/ee3b5aa3-2153-4b00-acf4-60bd64790eda" />

(How can it think I've misspelled "settings" when I've clearly written "SORT settings" grrrrr)

I just do this:

<img width="724" height="193" alt="Image" src="https://github.com/user-attachments/assets/b9faf602-908f-422e-b9cb-e0169aa2b4a1" />

And then rename the symbol afterwards. It's great. I love it. No notes.

**Comment by taeuscherpferd** [comment]
If the concern is workflow disruption, it feels like the proportion of normal users who this disrupts, is much higher than the proportion of those who one shot fly through their code accepting whatever change the IDE wants to give them without even looking at it. 

For them the workflow disruption is "Oh the IDE did something weird. Does that to me all the time. Oh well".

 But for the rest of us it's "Where's my button? I was certain that it was here somewhere. Is it in the right click menu? Maybe it's in the settings somewhere. (searches VSCode settings) Don't see it in here, did I get an update? Maybe there's a bug (Searches VSCode repo) Ah yes a [bug...](https://github.com/microsoft/vscode/issues/198682) oh it's closed, links to another bug. (Reads chat and is sad that there is no way to configure this to always be available and is now perpetually thinking about something they can never have. Takes time to write a silly comment in the hopes that one day we can have editing freedom.)

One disruption takes like 5m, the other like an hour. ¯\\\_(ツ)\_/¯

Hopefully we can at least get the option to keep it in the menu, would also be more than ok with it perpetually being in 2nd position 👍
